File: leetcode/golang/169_majority_element.go
func majorityElement(nums []int) int {
    res, sum := 0, 0
    for i := 0; i < len(nums); i++ {
        if sum == 0 {
            res = nums[i]
        }
        if res == nums[i] {
            sum += 1
        }
        if res != nums[i] {
            sum -= 1
        }
    }
    return res
}

File: leetcode/c++/169_majority_element.cpp
// https://leetcode.cn/problems/majority-element/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int res = nums[0], sum = 0;
        for (int num : nums) {
            if (sum == 0) 
                res = num;
            if (num == res) 
                sum += 1;
            else if (num != res)
                sum += -1;
        }
        return res;
    }
};

// 采用摩尔投票，时间复杂度为O(n)，空间复杂度为O(1)